aten/src/ATen/CPUApplyUtils.h

```diff
@@ -22,289 +24,224 @@ namespace at {
-\
-      if(ATENSOR##_counter[ATENSOR##_i]  == ATENSOR##_sizes[ATENSOR##_i]) \
-      { \
-        if(ATENSOR##_i == 0) \
-        { \
-          TH_TENSOR_APPLY_hasFinished = true; \
-          break; \
-        } \
-          else \
-        { \
-          /* Reset the pointer to the beginning of the chunk defined by this dimension */ \
-          ATENSOR##_data -= ATENSOR##_counter[ATENSOR##_i]*ATENSOR##_strides[ATENSOR##_i]; \
-          ATENSOR##_counter[ATENSOR##_i] = 0; \
-        } \
-      } \
-      else \
-        break; \
-    } \
-    ATENSOR##_i = 0; \
+// TODO:
+// Create a larger set of test cases for non-contiguous tensors
+// Optimize for parallelization and performance
+
+template <typename T>
+struct strided_tensor_iter {
+ private:
+  const bool is_contiguous_;
+  const int64_t dim_;
+  int64_t offset_;
+  const int64_t numel_;
+  std::vector<int64_t> counter_;
+  std::vector<int64_t> sizes_;
+  std::vector<int64_t> strides_;
+  T* data_ = NULL;
+  bool is_finished_;
+
+ public:
+  strided_tensor_iter(strided_tensor_iter const&) = delete;
+  void operator=(strided_tensor_iter const& x) = delete;
+  strided_tensor_iter(strided_tensor_iter&&) = default;
+  strided_tensor_iter(Tensor& tensor)
+      : is_contiguous_(tensor.is_contiguous()),
+        dim_(tensor.ndimension()),
+        offset_(0),
+        numel_(tensor.numel()),
+        data_(tensor.data<T>()),
+        is_finished_(false) {
+    if (!is_contiguous_) {
+      counter_.resize(dim_, 0);
+      sizes_.resize(dim_, 1);
+      strides_.resize(dim_, 1);
+      for (int64_t i = 0; i < dim_; i++) {
+        sizes_[i] = tensor.size(i);
+        strides_[i] = tensor.stride(i);
+      }
+    }
   }
-
-template <typename scalar1, typename scalar2, typename Op>
-void CPU_tensor_apply2_dim(Tensor& tensor1, Tensor& tensor2, int64_t dim, Op op) {
-  checkBackend("CPU_tensor_apply2", {tensor1, tensor2}, Backend::CPU);
-  bool TH_TENSOR_APPLY_hasFinished = false;
-  int64_t TH_TENSOR_dim_index = 0;
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar1, tensor1, dim, 1)
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar2, tensor2, dim, 1)
-  auto t1_numel = tensor1.numel();
-  auto t2_numel = tensor2.numel();
-  if(t1_numel != t2_numel) {
-    std::ostringstream oss;
-    oss << "inconsistent tensor size, expected " << tensor1.sizes() << " and " << tensor2.sizes()
-        << " to have the same number of elements, but got " << t1_numel << " and " << t2_numel << " elements respectively";
-    throw std::runtime_error(oss.str());
+  inline T* data() {
+    return data_ + offset_;
+  }
+  inline bool is_finished() {
+    return is_finished_;
   }
-  while(!TH_TENSOR_APPLY_hasFinished)
-  {
-    /* Loop through the inner most region of the Tensor */
-    for(; tensor1_i < tensor1_size && tensor2_i < tensor2_size; tensor1_i++, tensor2_i++, tensor1_data += tensor1_stride, tensor2_data += tensor2_stride)
-    {
-      op(*tensor1_data, *tensor2_data);
+  inline void iterate() {
```

Comment:
All struct methods are `inline` by default, but it has nothing to do with actual function inlining (only affects linkage).